coi/src/context.rs

```diff
@@ -73,20 +76,19 @@ impl UserInterests {
             compute_score_for_closest_positive_coi(embedding, &self.positive, horizon, time),
             compute_score_for_closest_negative_coi(embedding, &self.negative, horizon, time),
         ) {
-            (Some(positive), Some(negative)) => Some(positive - negative),
-            (Some(positive), None) => Some(positive),
-            (None, Some(negative)) => Some(-negative),
+            (Some(positive), Some(negative)) => Some((positive - negative + 2.) / 6.),
+            (Some(positive), None) => Some((positive + 1.) / 4.),
+            (None, Some(negative)) => Some((-negative + 1.) / 2.),
```

Comment:
> 🔧 The normalization for the compound score doesn't seem right shouldn't it be ( (p+1)/4 + (-n+1)/2 )/2?

if we normalize the positive and negative scores separately before combining them, then this would give equal weight to both of them, which is likely not what's intended. normalizing after the combination keeps their relative ratio.

> ❓ Wouldn't it be much less error prone (also wrt. to future changes) to do the normalization in the computation functions. 🤔 Through then the +1 part isn't really "just" normalization (but also shifting) so probably just normalizing both to [-1;1] in their function would be the best.

since we use an affine transformation (shifting and scaling) the original properties of the score are kept, it doesn't make a difference in that regard if we transfrom to `[-1,1]` or `[0,1]`. as we agreed some time ago to normalize scores to `[0,1]` i would like to handle it the same here.

> ❓ Does what we do here even make mathematically really make sense? Normalizing to [-1;1] only really makes sense if the distributions we compare assign a similar meaning to the same values in that range and gaps have the same meaning etc. But this isn't really the case here. Sure we do need to do some normalization for combining rankings from different sources, but then we should do the normalization there, not in this code.

i agree that it is difficult to assign some meaning to the normalized score of "positive score - negative score". it shouldn't be much of an issue though as we will likely remove the negative cois all together soon.

the issue of combining multiple normalized scores and their different meaning/distribution/etc isn't limited to this particular case, but more a general question. currently we deal with that by assigning different weights in the score combination. we could also think about more sophisticated transformations, but i would guess that it wouldn't improve the overall result by much, it seems much more important to have better models/embeddings on which we compute the scores in the first place.

Reply:
The first question point was more about weather it's best to normalize in the production functions instead of where we do now. Through with your point about why combining them before normalizing I guess that won't work.

Through we could give the negative CoI computation also a relevance value, but e.g. fix it to an specific value (e.g. 0)  and then normalize them separately  but in the same way (e.g. pretent that the negative range is also `[-1;3]`).

I'm aware that the shift by itself doesn't change the distribution. But if you then combine shifted distributions and normalize them again it can make a difference.

Mainly I'm worried a bit because the difference between <0, >0  in the original distribution has semantic meaning while we treat it as not having any meaning after normalization.

Reply:
> yes but if the normalization is quite specific to how we later in the pipeline want to combine things it would be more flexible to normalize later in the pipe, similar if we e.g. want to apply flexible weights provided as hidden endpoint parameters it would be simpler, if the normalization and any form of combination of scores is done later in the pipe

i agree to a cerain extend, the question is if we really need that much flexibility. it is usually sufficient to pass on normalized scores to a combinator function as long as the normalized score still contains the distribution information.

we would also need to expose internal details of the score range across the `coi` crate boundary, this can be error prone for multiple reasons: if internal scoring changes than the consumer crate has to be updated to account for the new unnormalized score, the docs have to be kept upt to date with the code, etc.

> Mainly I'm worried a bit because the difference between <0, >0 in the original distribution has semantic meaning while we treat it as not having any meaning after normalization.

this is already not given anymore, since for example the score for positive cois is computed as `similarity * decay + relevance`, so there is not break-even point (not even one different to zero) which has the meaning of changing from negative to positive or the like. all we can say about the score is that lower numbers mean less similar & lower relevance, whereas higher numbers mean more similar & higher relevance. so if we use a score like `[-1,1]` we would pretend some meaning is around 0, which there actually isn't. hence a normalized score in `[0,1]` seems most fitting.